Derivative related modifications

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
fixes #7027
closes #10597
closes #11962 
fixes #14719 
fixes #15028 
fixes #15084 
closes #13166 
fixes #13873
fixes #15131
fixes #6681
fixes #13457
fixes #15190 
fixes #6403
fixes #10150
closes #14221
fixes #15266

#### Brief description of what is fixed or changed

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Derivative has faster quick exit for 0
  * subs involving changing UndefinedFunction to Function in Derivative wrt variables will now create a Subs instance
  * error message for differentiation wrt bad variable no longer raised if the count is 0 and no longer refers to the count
  * error message for differentiation wrt variable that has a negative count
  * Derivative has `canonical` property which returns Derivative with `variable_count` canonically sorted
  * Derivative.variables is not a reliable way to see which are the variables of differentiation: it gives an expanded list of those variables, e.g `f(x,y).diff(x,2,y).variables -> (x,x,y)` or raises an error if some order of   differentiation is symbolic, e.g. `f(x).diff(x,y).variables -> TypeError`
  * Derivative has `_wrt_variables` which lists, in order, the variables of differentiation, even if the count is symbolic.
  * defined function and non-elementary Derivatives (like `Derivative(f(g(x)), x)`) give False for _diff_wrt
  * subs involving Subs now give a result that should be the same whether doit is applied before or after the subs, e.g. `Subs(x*y, x, x + 1).subs(x, y).doit() == Subs(x*y, x, x + 1).doit().subs(x, y)`
  * a change of symbols is now possible with Subs so `Subs(x, x, 1).subs(x, y) -> Subs(y, y, 1)`
  * `Function.fdiff` returns Subs less often, e.g. `f(g(x)).diff(x)` no longer creates a Subs instance
  * Expr has a default attribute `is_scalar=True` which makes Expr().diff(Expr()) == 1
  * bug involving derivatives of non-commutative expressions is fixed
  * Mul now implements the nth derivative in terms of multinomials
  * piecewise_fold is canonical wrt ordering of final Piecewise arguments
  * Piecewise uses information from Eq or And conditions to simplify the corresponding expression, e.g. (f(x), Eq(x, 0)) -> (f(0), Eq(x, 0))
  * BooleanFunction has `simplify` method
  * And will simplify args of type `Eq` when simple opportunities are presented, e.g. And(Eq(x, 0), Eq(x, y)) -> And(Eq(x, 0), Eq(y, 0))
* diffgeom
  * better removal of dummies from expressions involving derivatives
<!-- END RELEASE NOTES -->

- [x] check #14719 -- made a note on the issue page describing how to handle this

<!-- null -->
:100: All lines shown as added in the diff have nominal coverage.

smichr commented:
This will handle the recursion error:

```python
old = [log(a), log(-b)]
new = list(expand_log, old))
if new == old:
    return unsolved_result
return _solveset(new[0] - new[1], symbol, domain)
```

You can't use Piecewise to return a result because solveset can only deal with sets. It does not like a Piecewise. Given this, I don't see any point in keeping track of conditions.

smichr commented:
ping @asmeurer : this has removed differentiation wrt defined functions

asmeurer commented on sympy/core/function.py:
```diff
@@ -944,93 +953,45 @@ class Derivative(Expr):
     Symbol is attempted, the non-Symbol is temporarily converted to a Symbol
-    while the differentiation is performed.
-
-    Note that this may seem strange, that Derivative allows things like
```
I think this text should be kept, except with just talking about g(x) instead of cos(x)

asmeurer commented on sympy/core/function.py:
```diff
@@ -1200,9 +1177,20 @@ def __new__(cls, expr, *variables, **kwargs):
         # expression at all. Note, this cannot check non-symbols like
         # functions and Derivatives as those can be created by intermediate
         # derivatives.
-        if evaluate and all(isinstance(sc[0], Symbol) for sc in variable_count):
```
I'm not sure about this change. The motivation in the past is that for non-Symbol variables, we can't automatically assume that certain derivatives are 0. For example A[i].diff(A[j]) is KroneckerDelta(i, j), where `A = IndexedBase('A')`. Similar things happen with matrices. Granted, those things should all be tested, so is this still handling those cases properly?

smichr replied:
in this PR:
```
>>> A[i].diff(A[j])
KroneckerDelta(i, j)
```
This isn't 0 because these objects share a free symbol:
```
>>> A[i].free_symbols & A[j].free_symbols
{A}
```

asmeurer commented on sympy/core/function.py:
```diff
@@ -1200,9 +1177,20 @@ def __new__(cls, expr, *variables, **kwargs):
+            free = expr.free_symbols
+            for v, c in variable_count:
+                vfree = v.free_symbols
+                if c.is_positive and vfree:
```
I guess this was already there, but why is is_posistive needed here?

smichr replied:
A count can be positive, negative, 0 or symbolic (unknown). I am not sure what a negative power means -- antiderivative? I think the rationale here is we only know that differentiation wrt a quantity that has no symbol in common with the expr being differentiated if the count is positive. If the count is negative, I don't know what that means and if the count is 0 it doesn't matter. If the count is symbolic we don't know since it may be zero.

asmeurer replied:
Oh it's the count. Yes I remember this discussion now. It has to do with symbolic counts.

asmeurer commented on sympy/core/function.py:
```diff
@@ -922,8 +929,10 @@ class Derivative(Expr):
-        >>> from sympy.abc import x
+        >>> from sympy import cos, sin, sqrt, diff, Function
+        >>> from sympy.abc import x, y, z
+        >>> f, g = map(Function, 'fg')
```
You should use `symbols(cls=Function)` in docstrings.

smichr replied:
done

asmeurer commented:
Let's wait a bit to see if @Upabjojr has any comments.

smichr commented:
@asmeurer , take a look at the docstring of Derivative. I think I have kept the essentials of the original while cutting out a lot that either no longer applies or that can be demonstrated with more brevity.

smichr commented on sympy/core/function.py:
```diff
@@ -1257,8 +1257,6 @@ def __new__(cls, expr, *variables, **kwargs):
                     old_v = v
                     v = Dummy('xi_%i' % len(subs), dummy_index=hash(v))
                     expr = expr.xreplace({old_v: v})
-                    if not expr.has(v):
```
It seems that some sort of trickery is going on with vector objects. If you run this docstring test and watch expr you will see that it has a `u1` printed but the replace doesn't `u1(t)` with the dummy...so it looks like the answer should be 0 instead of `A.x`:

```python
>>> from sympy.physics.vector import ReferenceFrame, dynamicsymbols
>>> N = ReferenceFrame('N')
>>> A = ReferenceFrame('A')
>>> u1, u2 = dynamicsymbols('u1, u2')
>>> A.set_ang_vel(N, u1 * A.x + u2 * N.y)
>>> A.partial_velocity(N, u1)
A.x
```

I suspect that a "dynamicsymbol" has an explicit time dependence that its `_eval_derivative` knows about, 

This whole routine seems more complicated than it should be in keeping track of how different objects should be handled...

smichr commented:
@moorepants , could you see if there is any (further) degredation to performance as a result of these changes?

asmeurer commented:
You'll probably need to edit the wiki after this is merged to remove the other authors from the changes that were only you.

maurogaravello commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -2181,21 +2182,24 @@ def test_U4():
+    assert T.diff(x) == (
+        h(x)*Derivative(g(x), x)*Integral(f(y), y) +
+        g(x)*Derivative(h(x), x)*Integral(f(y), y))
 
```
The derivative should be `f(g(x)) * Derivative(g(x), x) - f(h(x)) * Derivative(h(x), x)`

jksuom commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -2181,21 +2182,24 @@ def test_U4():
+        Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2)
+    a2 = (
+        Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
+        Derivative(g(t), t)*Derivative(f(g(t)), g(t), t))
```
This doesn't look right. I don't think there should be a derivative wrt g(t).

smichr replied:
Here is my hand derivation:

```
f(g(t)).diff(t)

Derivative(f(g(t)),g(t))*Derivative(g(t),t)

identify as a*b, then (a*b).diff(t) is

Derivative(a, t)*b + a*Derivative(b, t)

Substituting a and b in derivatives

Derivative(Derivative(f(g(t)),g(t)), t)*b + a*Derivative(Derivative(g(t),t), t)

and simplifying

Derivative(f(g(t)), g(t), t)*b + a*Derivative(g(t), t, 2)

then substituting in a and b for the remaining occurrences gives a2

Derivative(f(g(t)), g(t), t)*Derivative(g(t),t) + Derivative(f(g(t)), g(t))*Derivative(g(t), t, 2)
\--------------------------------------------/

The first terms differ in a2 and a1 (below)

/---------------------------------------------------\
Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2 + Derivative(f(g(t)), g(t))*Derivative(g(t), t, 2)

Checking if the indicated expressions are plausibly equal

>>> f=lambda x:x**4
... g=lambda x:exp(x)
... 

from a2

>>> Subs(Derivative(f(x), x), x, g(t)).diff(t)*Derivative(g(t),t).doit()
12*exp(4*t)

from a1

>>> (Subs(Derivative(f(x),x,2),x,g(t))*(g(t).diff(t))**2).doit()
12*exp(4*t)

Now check all of a2

>>> (Subs(Derivative(f(x), x), x, g(t)).diff(t)*Derivative(g(t),t) + Subs(Derivative(f(x), x),x,g(t))*Derivative(g(t), t, 2)).doit()
16*exp(4*t)

and compare to original

>>> f(g(t)).diff(t,2)
16*exp(4*t)
```

jksuom replied:
a = Derivative(f(g(t)), g(t)) is short for Derivative(f(y), y).subs(y, g(t)), or Df(g(t)) if Df(y) is used to denote Derivative(f(y), y).

Then Derivative(a, t) is Derivative(Df(g(t)), g(t)))*Derivative(g(t), t) where
Derivative(Df(g(t)), g(t)) means Derivative(Df(y), y).subs(y, g(t)) which is
Derivative(f(y), (y, 2)).subs(y, g(t)).

I am not sure what Derivative(f(g(t)), g(t), t)) should mean but it looks different.

smichr replied:
Perhaps it's better to prefer that the evaluated form be represented with Subs:

```
>>> Derivative(f(g(x)),g(x))
Derivative(f(g(x)), g(x))
>>> _.doit()
Subs(Derivative(f(_xi_0), _xi_0), _xi_0, g(x))

Here's the expression in question:

>>> f(g(x)).diff(x,2)
Derivative(g(x), x)**2*Subs(Derivative(f(_xi_1), (_xi_1, 2)), _xi_1,
g(x)) + Derivative(g(x), (x, 2))*Subs(Derivative(f(_xi_1), _xi_1),
_xi_1, g(x))

A particular case of f and g:

>>> f(g(x)).diff(x,2).subs(f,Lambda(x,x**4)).subs(g,Lambda(x,exp(x))).doit()
16*exp(4*x)
```

smichr replied:
> I am not sure what Derivative(f(g(t)), g(t), t)) should mean but it looks different.

Are you mainly objecting to this representation? It means what you have said: `Subs(f(y).diff(y), y, g(t)).diff(t)`. I *think* that `Derivative(f(g(t)), g(t), t)).doit()` will spell this out explicitly but will check on this later.

jksuom replied:
> Subs(f(g(t)).diff(t), g(t), y) is 0 since you are differentiating wrt t.

This is not necessarily 0, if `g(t)` depends on `t`. For example, if `f(y) = y**2`, then `f(g(t)).diff(t)` is `2*g(t)*g(t).diff(t)` which becomes `2*y*g(t).diff(t)`.
This is the reason why I would like to ban derivative expressions having both `g(t)` and `t`.

smichr replied:
Here's why I think it's always zero:

```python
>>> f(g(t)).diff(t)
Derivative(g(t), t)*Subs(Derivative(f(_xi_1), _xi_1), (_xi_1,), (g(t),))
```
When you replace g(t) with y that first factor will be 0 and that forces the whole expression to 0. This is true for your example, too, since `Subs(g(t).diff(t), g(t), y)` will be 0.

I'm very keen to get this right, however, and we are already doing something that Ronan had warned against but others felt that the "function as symbol" semantics for derivation was important sugar to keep. So please continue to point out notation that you think is ambiguous. I am especially interested to know if there is actually a case where the order of differentiation matters. I think for our current assumptions that it does not.

jksuom replied:
> When you replace g(t) with y that first factor will be 0

I think that g(t) cannot be replaced *inside a derivative*. That is bound to give wrong results.

smichr replied:
I'm not sure what meaning `Subs(g(t).diff(t), g(t), y)` can have: `g(t).diff(t)` no longer depends on `g(t)`. This is the current assumption of derivatives wrt functions: `f(x).diff(x,f(x)) = 0`. Derivatives of the same expression to different orders are independent of each other hence f(x) (and order 0 derivative) and f(x).diff(x) (a first derivative) are independent. This is the assumption on which derivatives in SymPy are defined.

smichr replied:
This is the note that appears in Derivative's docstring:

```
    When differentiation wrt a non-Symbol is attempted, the non-Symbol
    is temporarily converted to a Symbol while the differentiation
    is performed::

        >>> eq = x + f(x)**2
        >>> eq.diff(f(x)).subs(f(x), y)
        2*y
        >>> eq.subs(f(x), y).diff(y)
        2*y

    This means, also, that all functions are treated as independent
    of other functions and are not related to their free symbols::

        >>> x.diff(f(x))
        0
        >>> g(x).diff(f(x))
        0

    It also means that derivatives of functions are assumed to be
    independent of the functions themselves and derivatives of
    different orders::

        >>> from sympy import Derivative
        >>> F = f(x)
        >>> dF = Derivative(F, x)
        >>> d2F = Derivative(F, x, 2)
        >>> F.diff(dF)
        0
        >>> dF.diff(F)
        0
        >>> dF.diff(d2F)
        0
```

jksuom replied:
It seems that SymPy will evaluate `a1` and `a2` in the same way.
```
>>> from sympy.abc import t
>>> f, g = symbols('f g', cls=Function)
>>> a1 = (
...         Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
...         Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2)
>>> a2 = (
...         Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
...         Derivative(g(t), t)*Derivative(f(g(t)), g(t), t))
>>> a1.doit() == a2.doit()
True
```
So I take back my doubts and agree that both `a1` and `a2` are correct. (I should have done this test in the first place, sorry.)

However, I also get
```
>>> f(g(t)).diff(t,2) in (a1, a2)
False
>>> f(g(t)).diff(t,2).doit() == a1.doit()
True
```
Perhaps we could dispense with `a2`?

jksuom replied:
Those were the results from master. In this branch I get `a2`.
```
>>> f(g(t)).diff(t, 2)
Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) + Derivative(g(t), t)*Derivative(f(g(t)), g(t), t)
```
Would it be hard to make this `a1`? (Automatically processing higher derivatives wrt both `g(t)` and `t`?)

smichr replied:
The reason it doesn't get processed is because all processing past the first non-symbol is halted. I asked on the mailing list for examples where the ordering of g(t) and t matters. Aaron has said that with our assumptions it doesn't matter. So I will modify the sorting routine to move all symbols to the front of the queue. That should take care of this.

smichr commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -2181,21 +2182,25 @@ def test_U4():
-    # raises ValueError: Invalid limits given: (y, h(x), g(x))
     T = integrate(f(y), y, h(x), g(x))
-    T.diff(x)
+    assert T.diff(x) == (
```
I think this should be
```
f(g(x))*g(x).diff(x)-f(h(x))*h(x).diff(x)
```

jksuom replied:
That seems right to me.

smichr replied:
The previous or the new? If you look at the `Integral._eval_derivative` it is more complicated...but it for sure did not get it right as it is because `y` is still in the answer that was given for the U6 result. I have to think about how this is modified with multiple integrations.

jksuom replied:
Sorry, I should have said that I agree. What you propose looks right to me.

asmeurer commented:
> Expr has a default attribute is_scalar=True which makes Expr().diff(Expr()) == 1

What is the motivation for this? It seems to me that Expr._diff_wrt should default to False, so that derivatives wrt Expr subclasses shouldn't even be allowed unless explicitly enabled.

asmeurer commented:
Or do you mean `ExprSubclass().diff(ExprSubclass()) == 1` *if `ExprSubclass._diff_wrt == True`*?

asmeurer commented:
What's the status of this? Should we try to get it in for 1.3 (I want to start the branch next week), or should we hold it off for the next release? I'm hoping with the auto-release notes that we will be able to release much more often, multiple times a year ideally. The 1.3 release will be a test of this hypothesis.

smichr commented:
> What's the status of this?

I am fixing the wester U6 calculation (i.e. derivative of Integral). Everything else has been gone over well by myself but needs someone else to review it.

smichr commented:
> fixing the wester U6 calculation

whew. No problem. The test was not written in the correct format to indicate upper and lower limits.

smichr commented:
> What is the motivation for this?

The motivation is issue #14719 

>  It seems to me that Expr._diff_wrt should default to False, so that derivatives wrt Expr subclasses shouldn't even be allowed unless explicitly enabled.

I think this is what you mean (and how it works in this branch):
```python
>>> class foo(Expr):
...  pass
...
>>> x.diff(foo())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\core\expr.py", line 3039, in diff
    return Derivative(self, *new_symbols, **assumptions)
  File "sympy\core\function.py", line 1187, in __new__
    __)))
ValueError:
Can't calculate derivative wrt foo().
```

smichr commented:
I would really like to have a graceful way to fix the negative order of differentiation but I don't know a clever way to get around the issue that series brings up in its use of the multinomial as discussed in #13873.

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -1007,7 +1030,7 @@ def piecewise_fold(expr):
         # (and the expression is commutative).
         if expr.is_Add or expr.is_Mul and expr.is_commutative:
-            p, args = sift(expr.args, lambda x: x.is_Piecewise, binary=True)
+            p, args = sift(list(ordered(expr.args)), lambda x: x.is_Piecewise, binary=True)
```
no, you can't arbitrarily sort Piecewise args.

smichr commented on sympy/core/tests/test_function.py:
```diff
@@ -965,3 +998,72 @@ def test_undef_fcn_float_issue_6938():
+    eq = f(x, g(x))
+    assert eq.diff((g(x), y)) == Derivative(f(x, g(x)), (g(x), y))
+    # issue 13166
+    assert eq.diff(x, 2) == (  # XXX check
```
this needs to be checked

smichr commented:
@jksuom , do you think the expression `Subs(Derivative(f(y, g(x)), y, x), y, x)` might better be shown as 
`Derivative(g(x), x)*Subs(Derivative(f(x, y), x, y), y, g(x))`?

jksuom commented:
I think that `Derivative(g(x), x)*Subs(Derivative(f(x, y), x, y), y, g(x))` would be preferable especially if there is going to be further processing. `Subs(Derivative(f(y, g(x)), y, x), y, x)` takes less space but one has to be very careful when computing its derivatives.

smichr commented:
@asmeurer , I suspect getting someone to review this will take longer than you want to wait for the next release. I am currently testing the removal of `_eval_diff`.

smichr commented:
> one has to be very careful

I made a modification so the symbols will always come before the functions:
```python
>>> f(x, g(x)).diff(g(x), x)
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) +
Derivative(f(x, g(x)), x, g(x)) <-- ok
>>> _.doit()
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) +
Derivative(g(x), x)*Subs(Derivative(f(x, _xi_0), (_xi_0, 2)), _xi_0,
g(x)) + Derivative(f(x, g(x)), x, g(x)) <-- ok
```

jksuom commented:
I have understood that `Derivative(f(x, g(x)), (g(x), 2))` would be a shorthand notation for `Subs(Derivative(f(x, _xi_0), (_xi_0, 2)), _xi_0, g(x))`. However, they both appear here:
```
>>> _.doit()
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) +
Derivative(g(x), x)*Subs(Derivative(f(x, _xi_0), (_xi_0, 2)), _xi_0,
g(x)) + Derivative(f(x, g(x)), x, g(x)) <-- ok
```
Is that correct?

smichr commented:
> Is that correct?

Yes.

moorepants commented:
> @moorepants , could you see if there is any (further) degredation to performance as a result of these changes?

You can run the benchmarks for before and after on your branch to see if anything arises. I only have things setup to run benchmarks after they are merged into master.

jksuom commented:
It seems to me that there is something wrong with `doit()`. I have understood that it can change the form of an expression but not its value.
```
>>> Derivative(f(x, g(x)), x, g(x))
Derivative(f(x, g(x)), x, g(x))
>>> _.doit()
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) + Derivative(f(x, g(x)), x, g(x))
>>> _.doit()
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) + Derivative(g(x), x)*Subs(Derivative(f(x, _xi_0), (_xi_0, 2)), _xi_0, g(x)) + Derivative(f(x, g(x)), x, g(x))
>>> _.doit()
Derivative(f(x, g(x)), (g(x), 2))*Derivative(g(x), x) + 2*Derivative(g(x), x)*Subs(Derivative(f(x, _xi_0), (_xi_0, 2)), _xi_0, g(x)) + Derivative(f(x, g(x)), x, g(x))
```

smichr commented:
> something wrong with doit()

Funny you should mention this! No kidding, I just noticed that in testing before I checked here. I know what the problem is and am working on it. Things are actually starting to look pretty sound.

smichr commented on sympy/core/function.py:
```diff
@@ -1230,13 +1230,26 @@ def __new__(cls, expr, *variables, **kwargs):
+                        denest = False
+                        break
+                if not denest:
+                    break
```
Maybe this should be
```
if denest:
    variable_count = list(expr.variable_count) + variable_count
    expr = expr.expr
else:
    return expr.func(expr.expr, *variable_count, **kwargs)
```

The first correction of U5 was better than the last one. What should `Derivative(f(g(x),h(x)).diff(g(x),h(x)).diff(x)` give?

smichr replied:
U5 is back to its compact form

smichr committed:
Piecewise._eval_simplify is defined

An attempt is made to use information in an Eq that is a
condition or is in an And that is a condition to simplify
the condition and the expression associated with it.

smichr committed:
diff(Integral does not return Integral(0

smichr committed:
modify doit methods

test for issue #15266

smichr committed:
add Derivative._wrt_variables

smichr committed:
modify Derivative._eval_subs and fdiff

* Derivative.canonical also added

smichr committed:
differentiation modifications

Notes:

Objects define their own diff if they want to take control of how
differentiation is handled: diff(foo) will call foo.diff() if it is
defined. There is no need for `_eval_diff` routines (though the
definitions of diff could be renamed _diff like subs. In fact, it
might be better to do so: canonicalization and sanitation of input
happens in diff and then _diff is called once something significant
needs to take place.

Here is the subs model

.subs - handles sanitizing of input and calls _subs if there is
        something to be done
    ._subs - handles traversal of arguments
        ._eval_subs - handles substitution of old->new for an object

diff doesn't have as clearly delineated operations.
It would be good to split the sanitization of input and maybe trivial zero
detection from the process of calculation the sequence of derivatives.
Methods that define diff methods should probably be defining _diff
methods.

.diff - handles sanitizing and trivial zero detection and
        calls _eval_diff routines
    ._eval_diff - called from diff to compute d_/dx
    ._eval_diff_n_times - called from diff to compute d^n_/dx^n
.fdiff - computes the 1st derivative of a functions ith argument

smichr commented:
check on relationship to #14221 and #10150

smichr committed:
watch for no order term

smichr commented:
:100: nominal coverage of modified lines is complete

smichr committed:
boolalg gets _eval_simplify method

Eq and Ne get their own simplification methods, too.
Simplifications of Ne do not propogate, e.g. Ne(x,1) & Eq(x, y) does
not imply Eq(y, 1) it implies Ne(y, 1) & Eq(x, y) but this additional
condition is not added to the args being processed.

asmeurer commented:
Is `_eval_derivative` deprecated here?

asmeurer commented:
I don't think this addresses https://github.com/sympy/sympy/issues/11568, but it's worth being aware of that discussion.

smichr commented:
> I don't think this addresses

Can you give an example of something that is not working?

smichr commented:
> Is `_eval_derivative` deprecated here?

That's a private method...does it need deprecation? It is removed but not deprecated. See comments, in the commit making that change, about the motivation.

asmeurer commented:
It's a method that subclasses might use to implement derivatives, including user-created subclasses. So I'd say it's public API. Does it still work in this branch if a subclass implements `_eval_derivative`? 

Regarding issue #11568, I will need to catch up on the discussions, but from what I saw of the code the unchecked box is not addressed here (not saying it should be). Basically, further decoupling of checking for Symbol and free_symbols is needed, so that hacks like the current `IndexedBase.free_symbols` are no longer needed. But like I said, it's just something to be aware of. I think it requires some more thought before we can proceed with any changes, especially if it will require a new API.

smichr commented:
> Does it still work in this branch if a subclass implements `_eval_derivative`?

No. But it's easy enough to revert that change.

asmeurer commented:
We might consider deprecating it instead of outright removing it. I consider any `_eval_*` method to be public API in the sense that custom subclasses are supposed to override those methods (the behavior of a specific `_eval_*` method in SymPy itself is not public API, but `diff` calling `_eval_derivative` is).

smichr commented:
I'm not going to push the structural change of `_eval_derivative` to `_eval_diff` because I'm not convinced that diff needs to be split up like subs. After sanitizing input and checking for quick exits, Derivative loops through the variables of differentiation and makes calls to the `_eval_derivative_n_times` routines and finally simplifies at the end.

smichr commented:
- [ ] rebase on master and fix usage of `linear_coeffs` in boolalg.

smichr commented:
If there are further changes needed or this acquires merge conflicts, please feel free to fork, fix and close this PR.

moorepants commented:
If this is ready, it should be merged soon. it will be hard for others to handle future merge conflicts.

moorepants commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -1958,15 +1958,7 @@ def test_diff():
-            [0,0],
-            [0,0]
-        ]]])
+    assert A.diff(B) == A.zeros(2)
```
Was this an incorrect result before?

smichr replied:
If something is detected as having a zero result it was returned like this:
https://github.com/sympy/sympy/blob/03ffa11604bcd59a75e3bb6acebe9e34416f8c65/sympy/core/function.py#L1273-L1277
The detection has been improved so it now gives this quick exit result. If it is important to keep the previous behavior then an additional `if` in the if-block will need to be added that tests whether a matrix is in the 'wrt' variables.

moorepants replied:
I'm not sure what the result is supposed to be. Is differentiating a 3x3 matrix wrt to a 2x1 matrix supposed to give a 2x1? Wouldn't it give a 3x3x2?

moorepants replied:
Previous result is a 2x1x2x2. Not sure where that comes from:
```
In [1]: import numpy as np

In [2]: a=np.array(
   ...:         [[[
   ...:             [0,0],
   ...:             [0,0]
   ...:         ]],
   ...:         [[
   ...:             [0,0],
   ...:             [0,0]
   ...:         ]]])
   ...:         

In [3]: a.shape
Out[3]: (2, 1, 2, 2)
```

smichr replied:
> Not sure where that comes from

nor I.

asmeurer replied:
CC @Upabjojr

Upabjojr replied:
I think the shape should be 2x1x3x3.

smichr replied:
> shape should be 2x1x3x3.

Please open a new issue if this is the case. I won't be adding more to this PR.

moorepants commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -361,7 +361,19 @@ def test_piecewise_simplify():
                   ((-1)**x*(-1), True))
     assert p.simplify() == \
-        Piecewise((1 + 1/x**2, Eq(x, 0)), ((-1)**(x + 1), True))
+        Piecewise((zoo, Eq(x, 0)), ((-1)**(x + 1), True))
```
Wrong result?

smichr replied:
simplification now substitutes known values of variables contained in the condition. `(1/x).subs(x,0)` gives `zoo` so `zoo` is now returned. Perhaps it would be better to leave the expression instead of the substituted form if any `NaN` or `zoo` is introduced?

moorepants replied:
That seems like a reasonable change. I guess we don't hold the outputs of simplify to any kind of strict solution, since it is impossible to define the simplest solution. No change needed. Just checking for potential backwards compatibility issues. Changed tests always raise flags for me.

moorepants commented on sympy/core/tests/test_function.py:
```diff
@@ -977,11 +1038,6 @@ def test_issue_13843():
 
-def test_issue_13873():
-    from sympy.abc import x
-    assert sin(x).diff((x, -1)).is_Derivative
```
Why remove this?

smichr replied:
Negative counts are no longer permitted; it is tested at line 1171 below:

https://github.com/sympy/sympy/blob/03ffa11604bcd59a75e3bb6acebe9e34416f8c65/sympy/core/tests/test_function.py#L1171-L1173

moorepants replied:
Ok, this is fine. Odd that the original test didn't test for an error being that it references the issue.

smichr replied:
It originally just left them unevaluated. Then when `series` generated negative indices and divided by `factorial(neg)` (which is `oo`) it eliminated the term having a Derivative with a negative count.

smichr committed:
closes #15386 by docstring note

smichr committed:
Function gets xreplace

smichr committed:
tighten up case (1) restriction on subs

test case that would fail without this change is added;
case for when there is a complete replacement of expr
is moved to outside the checking of a changed expr.

smichr commented:
From my perspective, this is ready for final review and merging. My only concern has been addressed in the last commit (and a test that would have failed in the previous form of the PR was added).

certik commented:
I reviewed the whole PR, it fixes a lot of issues, and all review comments were addressed. We can further improve upon it after it's merged.

Thank you @smichr !

smichr commented:
Thanks, @certik.

asmeurer commented:
Should "not a reliable way" have been "now a reliable way" in the release notes?
